pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/OpChainSchedulerService.java

```diff
@@ -31,151 +33,84 @@
-                LOGGER.debug("({}): Completed {}", operatorChain, operatorChain.getStats());
-              }
+              LOGGER.debug("({}): Completed {}", operatorChain, operatorChain.getStats());
             }
           } catch (Exception e) {
             LOGGER.error("({}): Failed to execute operator chain! {}", operatorChain, operatorChain.getStats(), e);
             thrown = e;
           } finally {
-            if (returnedErrorBlock || thrown != null) {
+            if (returnedErrorBlock != null || thrown != null) {
+              if (thrown == null) {
+                String blockMsg;
+                if (returnedErrorBlock.getDataBlock() instanceof MetadataBlock) {
+                  MetadataBlock metadataBlock = (MetadataBlock) returnedErrorBlock.getDataBlock();
+                  blockMsg = String.join(", ", metadataBlock.getExceptions().values());
+                } else {
+                  blockMsg = "Unknown";
+                }
+                thrown = new RuntimeException("Error block " + blockMsg);
+              }
               cancelOpChain(operatorChain, thrown);
             } else if (isFinished) {
               closeOpChain(operatorChain);
             }
           }
         }
       });
-    }
+    _submittedOpChainMap.put(operatorChain.getId(), scheduledFuture);
   }
 
-  /**
-   * Register a new operator chain with the scheduler.
-   *
-   * @param operatorChain the chain to register
-   */
-  public final void register(OpChain operatorChain) {
-    operatorChain.getStats().queued();
-    _scheduler.register(operatorChain);
-    LOGGER.debug("({}): Scheduler is now handling operator chain listening to mailboxes {}. "
-            + "There are a total of {} chains awaiting execution.", operatorChain,
-        operatorChain.getReceivingMailboxIds(),
-        _scheduler.size());
-  }
-
-  /**
-   * Async cancel a request. Request will not be fully cancelled until the next time opChain is being polled.
-   *
-   * @param requestId requestId to be cancelled.
-   */
-  public final void cancel(long requestId) {
-    _cancelledRequests.put(requestId, requestId);
-  }
-
-  /**
-   * This method should be called whenever data is available for an {@link OpChain} to consume.
-   * Implementations of this method should be idempotent, it may be called in the scenario that no data is available.
-   *
-   * @param opChainId the identifier of the operator chain
-   */
-  public final void onDataAvailable(OpChainId opChainId) {
-    _scheduler.onDataAvailable(opChainId);
-  }
-
-  // TODO: remove this method after we pipe down the proper executor pool to the v1 engine
-  public ExecutorService getWorkerPool() {
-    return _workerPool;
+  @Override
+  public void cancel(long requestId) {
+    // simple cancellation. for leaf stage this cannot be a dangling opchain b/c they will eventually be cleared up
+    // via query timeout.
+    List<OpChainId> opChainIdsToCancel = _submittedOpChainMap.keySet()
```

Comment:
Use `entrySet().iterator()` to iterate should be cheaper and create less garbage. We also need to remove the entry because there is no guarantee that the future already starts executing

Reply:
I've changed the code, but it is not clear to me the requirements of this method.

It is also not clear to me whether we can use `entrySet().iterator()` because we are going to be modifying the map to remove entries from the iterator while other threads will be adding elements. By the specification, the behavior there is undefined. Specifically, javadoc says we can remove elements from the entrySet, but doesn't specify what happens when other threads modify the map while iterating.

Given that this method is not going to be executed that often and the list shouldn't be that large, I don't think it is worth to enter in unspecified territory in order to do not create a short living array list.

Reply:
> We also need to remove the entry because there is no guarantee that the future already starts executing

In fact the entry was never removed, even if the future was executed, right?

Reply:
yes we need to clean up the map i think i commented in https://github.com/walterddr/pinot/pull/58#discussion_r1189345501

Reply:
Jackie suggested to remove the entry once the task finish and I modified the code to do so